protocol/consistencychecks.go

```diff
@@ -0,0 +1,298 @@
+	}
+
+	cc := &ConsistencyChecks{
+		CurrentEpoch: 0,
+		SavedSTR:     savedSTR,
+		useTBs:       useTBs,
+		signKey:      signKey,
+	}
+
+	if useTBs {
+		cc.TBs = make(map[string]*m.TemporaryBinding)
+	}
+	return cc
+}
+
+// Verify verifies the directory's response for a request.
+// It first verifies the directory's returned status code of the request.
+// If the status code is not in ErrorResponses array, it means
+// the directory has successfully handled the request.
+// The verifier will then verify the consistency state of the response,
+// and finally, it verifies the returned proof type
+// according to the request type.
+// This will also update the consistency state if the check is passed.
+func (cc *ConsistencyChecks) Verify(requestType int, msg *Response,
+	uname string, key []byte) ErrorCode {
+
+	if ErrorResponses[msg.Error] {
+		return msg.Error
+	}
+
+	var verifResult error
+	var proofType int
+
+	switch msg.DirectoryResponse.(type) {
+	case *DirectoryProof:
+		proofType, verifResult = cc.verifyDirectoryProof(
+			msg.DirectoryResponse.(*DirectoryProof), uname, key)
+	case *DirectoryProofs:
+		proofType, verifResult = cc.verifyDirectoryProofs(
+			msg.DirectoryResponse.(*DirectoryProofs), uname, key)
+	}
+
+	if verifResult != nil {
+		return verifResult.(ErrorCode)
+	}
+
+	switch requestType {
+	case RegistrationType:
+		if msg.Error == ErrorNameExisted &&
+			proofType == proofOfInclusion {
+			return PassedWithAProofOfInclusion
+		} else if msg.Error == Success &&
+			proofType == proofOfAbsence {
+			return PassedWithAProofOfAbsence
+		}
+		return cc.verifyProofTypeUsingPromises(msg.Error, proofType)
+
+	case KeyLookupType:
+		if msg.Error == ErrorNameNotFound &&
+			proofType == proofOfAbsence {
+			return PassedWithAProofOfAbsence
+		} else if msg.Error == Success &&
+			proofType == proofOfInclusion {
+			return PassedWithAProofOfInclusion
+		}
+		return cc.verifyProofTypeUsingPromises(msg.Error, proofType)
+
+	case MonitoringType:
+		if proofType == proofOfAbsence {
+			return ErrorBadProofType
+		}
+		return PassedWithAProofOfInclusion
+
+	case KeyLookupInEpochType:
+		// TODO:
+
+	}
+
+	// this return should be unreachable
+	// unless the developers mistakenly passed an invalid request type.
+	return ErrorMalformedClientMessage
+}
+
+func (cc *ConsistencyChecks) verifyDirectoryProof(df *DirectoryProof,
+	uname string, key []byte) (proofType int, verifResult error) {
+
+	verifResult = nil
+	proofType = invalidProof
+
+	ap := df.AP
+	str := df.STR
+	tb := df.TB
+
+	if verifResult = cc.verifySTR(str); verifResult != nil {
+		return
+	}
+
+	if proofType, verifResult = cc.verifyAuthPath(uname, key,
+		ap, str); verifResult != nil {
+		return
+	}
+
+	if verifResult = cc.verifyReturnedPromise(tb, uname, key, str, ap); verifResult != nil {
+		return
+	}
+
+	if verifResult = cc.verifyFulfilledPromise(uname, str, ap); verifResult != nil {
+		return
+	}
+
+	// re-assign new state for the client
+	if str.Epoch == cc.CurrentEpoch+1 {
```

Comment:
It's not obvious that a method called `verifyDirectoryProof` is going to be changing state.